examples/cluster/plot_cluster_comparison.py

```diff
@@ -3,116 +3,174 @@
         connectivity=connectivity)
+    spectral = cluster.SpectralClustering(
+        n_clusters=defaults['n_clusters'], eigen_solver='arpack',
+        affinity="nearest_neighbors")
+    dbscan = cluster.DBSCAN(eps=defaults['eps'])
+    affinity_propagation = cluster.AffinityPropagation(
+        damping=defaults['damping'], preference=defaults['preference'])
+    average_linkage = cluster.AgglomerativeClustering(
+        linkage="average", affinity="cityblock",
+        n_clusters=defaults['n_clusters'], connectivity=connectivity)
+    birch = cluster.Birch(n_clusters=defaults['n_clusters'])
+    gmm = mixture.GaussianMixture(
+        n_components=defaults['n_clusters'], covariance_type='full')
+
+    clustering_algorithms = (
+        ('MiniBatchKMeans', two_means),
+        ('AffinityPropagation', affinity_propagation),
+        ('MeanShift', ms),
+        ('SpectralClustering', spectral),
+        ('Ward', ward),
+        ('AgglomerativeClustering', average_linkage),
+        ('DBSCAN', dbscan),
+        ('Birch', birch),
+        ('GaussianMixture', gmm)
+    )
+
+    for name, algorithm in clustering_algorithms:
+        t0 = time.time()
 
-    birch = cluster.Birch(n_clusters=2)
-    clustering_algorithms = [
-        two_means, affinity_propagation, ms, spectral, ward, average_linkage,
-        dbscan, birch]
+        # catch warnings related to kneighbors_graph
+        with warnings.catch_warnings():
+            warnings.filterwarnings(
+                "ignore",
+                message="the number of connected components of the " +
+                "connectivity matrix is [0-9]{1,2}" +
+                " > 1. Completing it to avoid stopping the tree early.",
+                category=UserWarning)
+            warnings.filterwarnings(
+                "ignore",
+                message="Graph is not fully connected, spectral embedding" +
+                " may not work as expected.",
+                category=UserWarning)
+            algorithm.fit(X)
 
-    for name, algorithm in zip(clustering_names, clustering_algorithms):
-        # predict cluster memberships
-        t0 = time.time()
-        algorithm.fit(X)
         t1 = time.time()
         if hasattr(algorithm, 'labels_'):
             y_pred = algorithm.labels_.astype(np.int)
         else:
             y_pred = algorithm.predict(X)
 
-        # plot
-        plt.subplot(4, len(clustering_algorithms), plot_num)
+        plt.subplot(len(datasets), len(clustering_algorithms), plot_num)
         if i_dataset == 0:
             plt.title(name, size=18)
-        plt.scatter(X[:, 0], X[:, 1], color=colors[y_pred].tolist(), s=10)
-
-        if hasattr(algorithm, 'cluster_centers_'):
-            centers = algorithm.cluster_centers_
-            center_colors = colors[:len(centers)]
-            plt.scatter(centers[:, 0], centers[:, 1], s=100, c=center_colors)
-        plt.xlim(-2, 2)
-        plt.ylim(-2, 2)
+
+        colors = np.array(list(islice(cycle('bgrcmyk'),
```

Comment:
Use `['navy', 'yellowgreen', 'gold', 'cornflowerblue']` to be color-blind compatible.